File: libraries/KMPKtorCoinbase/shared/src/commonMain/kotlin/dev/subfly/kmpktorcoinbase/featureDetail/domain/usecase/GetCoinByIdUseCase.kt
package dev.subfly.kmpktorcoinbase.featureDetail.domain.usecase

import dev.subfly.kmpktorcoinbase.core.data.mapper.toCoinDetailModel
import dev.subfly.kmpktorcoinbase.core.exception.InvalidIdException
import dev.subfly.kmpktorcoinbase.core.model.CoinDetailModel
import dev.subfly.kmpktorcoinbase.core.util.Resource
import dev.subfly.kmpktorcoinbase.featureDetail.data.repository.CoinPaprikaDetailRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

class GetCoinByIdUseCase(private val repository: CoinPaprikaDetailRepository) {
    suspend fun fetchData(coinId: String?): Flow<Resource<CoinDetailModel>> = flow {
        emit(Resource.Loading)
        try {
            if (coinId == null) {
                emit(Resource.Error("Coin can not be found, please try again..."))
            } else {
                val detailFetched = repository
                    .getCoinDetail(id = coinId)
                    .toCoinDetailModel()
                emit(Resource.Success(detailFetched))
            }
        } catch (e: InvalidIdException) {
            emit(Resource.Error(e.message))
        } catch (e: Exception) {
            // General exception handling for Ktor, no deep dive, sorry
            e.message?.let { errorMessage ->
                emit(Resource.Error(errorMessage))
            }
        }
    }
}

File: ios/BaseiOSSwitUI/BaseiOSSwitUI/FeatureDetail/Domain/UseCase/GetCoinByIdUseCase.swift
//
//  GetCoinByIdUseCase.swift
//  BaseiOSSwitUI
//
//  Created by Ali Taha on 5.12.2023.
//

import Foundation
import Combine

class GetCoinByIdUseCase {
    private let repository: CoinPaprikaDetailDao
    
    var resource: CurrentValueSubject<Resource<CoinDetailModel>, Never>

    init(repository: CoinPaprikaDetailDao) {
        self.repository = repository
        self.resource = CurrentValueSubject(.loading)
    }

    func fetchData(coinId: String) async {
        do {
            resource.send(.loading)
            let coinFetched = try await repository.getCoinDetail(for: coinId).toCoinDetailModel()
            resource.send(.success(data: coinFetched))
        } catch let error as InvalidUrlException {
            resource.send(.error(message: error.message))
        } catch let error as InvalidIdException {
            resource.send(.error(message: error.message))
        } catch let error as InvalidResponseException {
            resource.send(.error(message: error.message))
        } catch {
            resource.send(.error(message: error.localizedDescription))
        }
    }
}
